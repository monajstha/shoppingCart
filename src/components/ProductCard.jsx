import React from "react";
import { redirect, useNavigate } from "react-router-dom";

export default function ProductCard({ productDetails }) {
  const navigate = useNavigate();
  const { id, title, image, price } = productDetails;

  const handleProductClick = () => {
    navigate(`/products/${id}`);
  };

  return (
    <div
      className="flex flex-col w-56 h-72  bg-white m-3 justify-center items-start cursor-pointer shadow-black shadow-lg hover:shadow-xl"
      aria-label={`View details for ${title}`}
      onClick={handleProductClick}
      role="button"
    >
      <div className="flex justify-center items-center w-full h-48 p-4 ">
        <img
          className="w-full h-40 object-contain"
          src={image}
          alt={title}
          onLoad={() => <>Loading..</>}
        />
      </div>
      <div className="p-2 bg-slate-700 w-[100%] text-lg h-full">
        <p className="line-clamp-2 text-lg">{title}</p>
        <p className="">${price}</p>
      </div>
    </div>
  );
}
