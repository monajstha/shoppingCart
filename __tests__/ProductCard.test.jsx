import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MemoryRouter } from "react-router-dom";
import ProductCard from "../src/components/ProductCard";

// Mocking useNavigate
const mockNavigate = vi.fn();

vi.mock("react-router-dom", async () => {
  const actual = await vi.importActual("react-router-dom"); // Keep the original functionality
  return {
    ...actual,
    useNavigate: () => mockNavigate, // Mock implementation
  };
});

describe("ProductCard", () => {
  beforeEach(() => {
    mockNavigate.mockClear(); // Clear the mock before each test
  });

  it("renders product details and navigates on click", async () => {
    const productDetails = {
      id: 1,
      title: "Sample Product",
      image: "https://via.placeholder.com/150",
      price: 29.99,
    };

    render(
      <MemoryRouter>
        <ProductCard productDetails={productDetails} />
      </MemoryRouter>
    );

    // Verify product details are rendered
    expect(screen.getByText(/Sample Product/i)).toBeInTheDocument();
    expect(screen.getByText(/\$29.99/)).toBeInTheDocument();
    expect(screen.getByAltText(/Sample Product/i)).toBeInTheDocument();

    // Simulate a click on the product card
    const card = screen.getByText(/Sample Product/i).closest("div");
    await userEvent.click(card);

    // Verify navigation was triggered
    expect(mockNavigate).toHaveBeenCalledWith(`/products/${productDetails.id}`);
  });
});

describe("ProductCard Snapshot", () => {
  it("matches snapshot", () => {
    const productDetails = {
      id: 1,
      title: "Sample Product",
      image: "https://via.placeholder.com/150",
      price: 29.99,
    };

    const { container } = render(
      <MemoryRouter>
        <ProductCard productDetails={productDetails} />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });
});
