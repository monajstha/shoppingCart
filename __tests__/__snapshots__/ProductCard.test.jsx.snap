// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductCard Snapshot > matches snapshot 1`] = `
<div>
  <div
    class="flex flex-col w-56 h-60  bg-slate-500 m-3 justify-center items-start cursor-pointer"
  >
    <img
      alt="Sample Product"
      class="w-56 h-36"
      src="https://via.placeholder.com/150"
    />
    <div
      class="p-2"
    >
      <p
        class="line-clamp-2"
      >
        Sample Product
      </p>
      <p>
        $
        29.99
      </p>
    </div>
  </div>
</div>
`;
